client/lib/webauthn/index.js

```diff
@@ -0,0 +1,186 @@
+/**
+ * Internal dependencies
+ */
+import wpcom from 'lib/wp';
+import { translate } from 'i18n-calypso';
+import config from 'config';
+
+let _backend;
+
+const POST = 'POST';
+
+export function strToBin( str ) {
+	str = str.replace( /[-_]/g, function( m ) {
+		return m[ 0 ] === '-' ? '+' : '/';
+	} );
+	return Uint8Array.from( atob( str ), c => c.charCodeAt( 0 ) );
+}
+
+export function binToStr( bin ) {
+	return btoa( new Uint8Array( bin ).reduce( ( s, byte ) => s + String.fromCharCode( byte ), '' ) );
+}
```

Comment:
This pair of functions seems very error prone, and some quick testing indicates that it has some issues. Are they expected to operate on a very limited set of inputs?

Drop this in a browser:

```js
function strToBin( str ) {
	str = str.replace( /[-_]/g, function( m ) {
		return m[ 0 ] === '-' ? '+' : '/';
	} );
	return Uint8Array.from( atob( str ), c => c.charCodeAt( 0 ) );
}
function binToStr( bin ) {
	return btoa( new Uint8Array( bin ).reduce( ( s, byte ) => s + String.fromCharCode( byte ), '' ) );
}

const inputs = [
    [ 'hex', '0123456789abcdef' ],
    [ 'simple', 'Hello, world!' ],
    [ 'complex', 'áæñü?+-_ ?=&' ],
    [ 'emoji', '😱😎\u00a0' ],
];

inputs.forEach( ( [ desc, input ] ) => {
  try {
    if ( input !== binToStr( strToBin( input ) ) ) {
      console.warn( '%o: %o did not match its conversion %o', desc, input, binToStr( strToBin( input ) ) )
    }
  } catch ( err ) {
    console.warn( 'Error in input %o: %o\n%o', desc, input, err.message );
  }
} )
// Error in input "simple": "Hello, world!"
// "Failed to execute 'atob' on 'Window': The string to be decoded is not correctly encoded."
// Error in input "complex": "áæñü?+-_ ?=&"
// "Failed to execute 'atob' on 'Window': The string to be decoded is not correctly encoded."
// Error in input "emoji": "😱😎 "
// "Failed to execute 'atob' on 'Window': The string to be decoded contains characters outside of the Latin1 range."
```

Should we add documentation, unit tests, and/or fixes and improvements? Can we use an external library?

Reply:
Thanks for the review! I think the function names should be probably renamed to `base64StringToBin/binStringToBase64` or something similar, `strToBin` is only expected to work with base64url encoded strings. We can likely use `base64-js` for `binToStr`.

Reply:
Looks like we can use https://www.npmjs.com/package/@github/webauthn-json, will do some tests and see if we can replace the current library.

Reply:
That sounds like a great idea @xknown do you want to try and replace the lib in this PR or should we try to do that in a subsequent PR?

Reply:
@xyu can you take a look at #36328? I think everything is working as expected, if you notice a problem we can likely postpone merging it to this PR.

Reply:
❤️ deleting code
```
+20 −160
```